lib/qless/middleware/timeout.rb

```diff
@@ -15,29 +16,45 @@ module Middleware
     # or a variable timeout based on the individual TTLs of each job
     # (using something like `extend Qless::Middleware::Timeout.new { |job| job.ttl * 1.1 }`).
     class Timeout < Module
-      def initialize
+      def initialize(opts = {})
+        timeout_class = opts.fetch(:timeout_class, ::Timeout)
+        kernel_class = opts.fetch(:kernel_class, Kernel)
         module_eval do
           define_method :around_perform do |job|
-            timeout_value = yield job
-            if timeout_value <= 0
-              raise InvalidTimeoutError, "Timeout must be positive, but was #{timeout_value}"
-            end
+            timeout_seconds = yield job
+            if timeout_seconds.nil?
+              super(job)
+            else
+              if !timeout_seconds.is_a?(Numeric) || timeout_seconds <= 0
+                raise InvalidTimeoutError, "Timeout must be a positive number or nil, " \
+                                           "but was #{timeout_seconds}"
+              end
 
-            begin
-              ::Timeout.timeout(timeout_value) { super(job) }
-            rescue ::Timeout::Error => e
-              error = JobTimedoutError.new(e.message)
-              error.set_backtrace(e.backtrace)
-              job.fail(*Qless.failure_formatter.format(job, error, []))
+              begin
+                timeout_class.timeout(timeout_seconds) { super(job) }
+              rescue ::Timeout::Error => e
+                error = JobTimedoutError.new(e.message)
+                error.set_backtrace(e.backtrace)
+                # The stalled connection to redis might be the cause of the timeout. We cannot rely
+                # on state of connection either (e.g., we might be in the middle of Redis call when
+                # timeout happend). To play it safe, we reconnect.
+                job.reconnect_to_redis
+                job.fail(*Qless.failure_formatter.format(job, error, []))
+                # Since we are leaving with bang (exit!), normal requeue logic does not work.
+                # Do it manually right here.
+                if self.is_a?(::Qless::Middleware::RequeueExceptions) && self.requeueable?(JobTimedoutError)
```

Comment:
It would seem cleaner if we could rely on the superposition of middlewares to handle retries here, instead of having the job check its own middlewares. Have you considered perhaps a middleware, like the RequeueMiddleware, that rescues exceptions coming through the middle ware stack and `kill!`s those that make it past the RequeueMiddleware?